command/agent/template/template.go

```diff
@@ -253,6 +265,20 @@ func newRunnerConfig(sc *ServerConfig, templates ctconfig.TemplateConfigs) (*ctc
 	// that need to be fixed for 1.7x/1.8
 	if sc.AgentConfig.Cache != nil && sc.AgentConfig.Cache.Persist != nil && len(sc.AgentConfig.Listeners) != 0 {
 		attempts = 0
+
+		// If we don't want exit on template retry failure (i.e. unlimited
+		// retries), let consul-template handle retry and backoff logic.
+		//
+		// Note: This is a fixed value (12) that ends up being a multiplier to
+		// retry.num_retires (i.e. 12 * N total retries per runner restart).
+		// Since we are performing retries indefinitely this base number helps
+		// prevent agent from spamming Vault if retry.num_retries is set to a
+		// low value by forcing exponential backoff to be high towards the end
+		// of retries during the process.
+		if sc.AgentConfig.TemplateConfig != nil && !sc.AgentConfig.TemplateConfig.ExitOnRetryFailure {
+			attempts = ctconfig.DefaultRetryAttempts
+		}
```

Comment:
This logic over here is what allows agent to perform a reasonable number of retries with enough backoff pause towards the tail end of the run before restarting. I've chosen `ctconfig.DefaultRetryAttempts` as the value since that was always the case on Agent 1.5.x and earlier. We can't completely hand off retries to the cache's proxy because the API client only performs retries on 5xx errors whereas the builtin retry logic from consul-template performs retries on any errors. A non-existent secret (i.e. a 404) without retry attempts from consul-template itself will spam apiproxy, and in turn the server, with requests indefinitely which can easily DDoS Vault.

The alternative here could be to set `attempts = sc.AgentConfig.Vault.Retry.NumRetries` but it has the danger of spamming the server with requests if the value is too low or zero due to low initial backoff values.